File: 011_Container_With_Most_Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left, right, res = 0, len(height) - 1, 0
        while left < right:
            h = min(height[left], height[right])
            res = max(res, (right - left) * h)
            while height[left] <= h and left < right:
                left += 1
            while height[right] <= h and left < right:
                right -= 1
        return res

File: 011_Container_With_Most_Water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0, right = height.size() - 1, res = 0;
        while (left < right) {
            int h = min(height[left], height[right]);
            res = max(res, h * (right - left));
            while (height[left] <= h && left < right) ++left;
            while (height[right] <= h && left < right) --right;
        }
        
        return res;
    }
};
